compiler/rustc_hir_analysis/src/check/wfcheck.rs

```diff
@@ -1755,20 +1755,41 @@ fn check_variances_for_type_defn<'tcx>(
             .collect::<FxHashSet<_>>()
     });
 
+    let ty_generics = tcx.generics_of(item.owner_id);
+
     for (index, _) in variances.iter().enumerate() {
         let parameter = Parameter(index as u32);
 
         if constrained_parameters.contains(&parameter) {
             continue;
         }
 
-        let param = &hir_generics.params[index];
+        let ty_param = &ty_generics.params[index];
+        let mut hir_param = &hir_generics.params[index];
+
+        if ty_param.name != hir_param.name.ident().name {
```

Comment:
```suggestion
        if ty_param.def_id != hir_param.def_id {
```
Working with def_ids is always more robust.